### PR TITLE
fix: properly convert osmosis legacy transactions

### PR DESCRIPTION
**Pull request by pastaghost**
* enforces snake_case variable naming convention
* adds support for single-sided-lp and bonding messages
* uses protobufs provided by osmojs
* makes `parse_legacy_tx_format` agnostic to `msg`/`msgs`

### CODE REVIEW

**Review by gomesalexandre — APPROVED** [review]
Untested but cross-checked against https://pkg.go.dev/github.com/osmosis-labs/osmosis/x/gamm/types / https://docs.osmosis.zone/osmosis-core/modules/gamm/ / https://docs.osmosis.zone/osmosis-core/modules/lockup/, LGTM

File: src/index.ts [hunk 1]
```diff
@@ -22,6 +22,18 @@ import BN from 'bn.js'
 import { osmosis, thorchain } from './amino'
 import * as codecs from './proto'
 
+type AgnosticStdTx = Omit<amino.StdTx, 'msg'> &
```
Comment (kaladinlight) [inline src/index.ts:25]: would probably be easier just to have a union of `amino.StdTx | StdTxWithMsgs` or something like that.

`export type StdTxWithMsgs = Omit<amino.StdTx, 'msg'> & { readonly msgs: readonly amino.AminoMsg[] }`

File: src/index.ts [hunk 2]
```diff
@@ -22,6 +22,18 @@ import BN from 'bn.js'
...
+  (
+    | {
+        readonly msg?: readonly amino.AminoMsg[]
```
Comment (kaladinlight) [inline src/index.ts:28]: `msg` is required for `amino.StdTx`

File: src/index.ts [hunk 3]
```diff
@@ -22,6 +22,18 @@ import BN from 'bn.js'
...
+        msgs?: never;
+      }
+    | {
+        readonly msgs?: readonly amino.AminoMsg[]
+        msg: never;
```
Comment (kaladinlight) [inline src/index.ts:33]: `msg` optional never (though, see above about different way to define this type that makes more sense to me)

File: src/index.ts [hunk 4]
```diff
@@ -22,6 +22,18 @@ import BN from 'bn.js'
...
 import * as codecs from './proto'
 
+type AgnosticStdTx = Omit<amino.StdTx, 'msg'> &
+  (
+    | {
+        readonly msg?: readonly amino.AminoMsg[]
+        msgs?: never;
+      }
+    | {
+        readonly msgs?: readonly amino.AminoMsg[]
```
Comment (kaladinlight) [inline src/index.ts:32]: `msgs` required for this case

File: src/index.ts (hunk 1 above)
Comment (kaladinlight) [inline src/index.ts:25]: `sign()` function sig should be updated to reflect new type

File: src/index.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/index.ts:32]: @pastaghost ☝🏽

File: src/index.ts [hunk 5]
```diff
@@ -122,31 +135,34 @@ const scrubCoin = (x: Coin) => {
...
+    tokenOutDenom: x.token_out_denom
   }
 }
 
 const scrubRoutes = (x: Route[]) => x.map(scrubRoute)
 
-function parse_legacy_tx_format({ fee, memo, msg, signatures }: amino.StdTx): ProtoTx {
-  if (msg.length !== 1) throw new Error('multiple msgs not supported!')
+function parse_legacy_tx_format(tx: AgnosticStdTx): ProtoTx {
+  const msgOrMsgs = tx.msg ?? tx.msgs
```
Comment (kaladinlight) [inline src/index.ts:153]: union type would allow proper narrowing here

Comment (kaladinlight) [reply src/index.ts:153]: `const getMessagesAmino = (tx: amino.StdTx | StdTxWithMsgs): readonly amino.AminoMsg[] => 'msgs' in tx ? tx.msgs : tx.msg`

File: src/index.ts [hunk 6]
```diff
@@ -306,52 +322,70 @@ function convertLegacyMsg(msg: amino.AminoMsg): Pick<ProtoTx, 'msg'> {
       }
     case 'osmosis/gamm/swap-exact-amount-in':
       if (!msg.value.sender) throw new Error('Missing sender in msg')
-      if (!msg.value.tokenIn) throw new Error('Missing tokenIn in msg')
-      if (!msg.value.tokenOutMinAmount) throw new Error('Missing tokenOutMinAmount in msg')
+      if (!msg.value.token_in) throw new Error('Missing token_in in msg')
```
Comment (kaladinlight) [inline src/index.ts:325]: seems like we could support both camel and snake case value keys if we wanted to

Comment (pastaghost) [reply src/index.ts:325]: We could support both if we wanted to, but I don't think we do. Unifying the case style across the stack is the way to go, IMO - having different tx formats in different places is a PITA and expands the surface for potential bugs.

**Comment by pastaghost** [comment]
@kaladinlight - I like the type union over introducing a new type here. The PR was merged as-is about half an hour ago - I'll open a new PR with the requested changes before propagating the changes from this PR upstream.

**Comment by pastaghost** [comment]
@kaladinlight @gomesalexandre - comments addressed in https://github.com/shapeshift/proto-tx-builder/pull/18

**Pull request merged**